stdpopsim/ext/selection.py

```diff
@@ -0,0 +1,211 @@
+        else:
+            raise ValueError(
+                    f"{self.distribution_type} is not a supported distribution type")
+
+
+def slim_mutation_frac(mutation_types):
+    """
+    The fraction of mutations that should be added by SLiM.
+    The remainder are added by msprime.mutate() once the SLiM part
+    of the simulation is complete.
+    """
+    if mutation_types is None:
+        weighted = False
+    else:
+        weighted = any(mut_type.weight > 0 for mut_type in mutation_types)
+    return 1 if weighted else 0
+
+
+@attr.s
+class GenerationAfter(object):
+    """
+    We often want to do things one generation after another event. This class
+    is essentially just a marker so that the SLiM engine can calculate the next
+    generation by taking into account the scaling parameter Q.
+    We call float(generation_after_obj) to get the time value.
+
+    XXX: Rename this? GenerationAfter means the generation after something
+         in a forwards-time convention, but times are specified with units of
+         generations before the present!
+    """
+    time = attr.ib(type=float)
+
+    def __attrs_post_init__(self):
+        validate_time(self)
+
+    def __float__(self):
+        return float(self.time)
+
+
+def validate_time(time):
+    if float(time) < 0:
+        raise ValueError("Negative times are invalid.")
+    if float(time) == 0 and isinstance(time, GenerationAfter):
+        raise ValueError("The generation after time=0 is in the future.")
+
+
+def validate_time_range(start_time, end_time):
+    validate_time(start_time)
+    validate_time(end_time)
+    if float(start_time) < float(end_time):
+        raise ValueError(f"start_time={start_time} < end_time={end_time})")
+    if (float(start_time) == float(end_time) and
+       isinstance(start_time, GenerationAfter) and
+       not isinstance(end_time, GenerationAfter)):
+        raise ValueError(
+                f"start_time=GenerationAfter({start_time}) < end_time={end_time}")
+
+
+class ExtendedEvent(object):
+    """
+    ExtendedEvent is analogous to msprime.DemographicEvent, but is used here
+    for SLiM things that msprime doesn't support. Times are all in units of
+    generations before the present, just like for msprime.DemographicEvent.
+    """
+    pass
+
+
+@attr.s(kw_only=True)
+class DrawMutation(ExtendedEvent):
+    """
+    TODO: docstring.
+
+    Draw a new mutation with the given mutation type in the given population
+    at the given genomic coordinate. The mutation will be added to one randomly
+    chosen chromosome in the given population.
+    If save=True, the simulation state is saved before the mutation is
+    introduced, to facilitate the save/restore mechanism used for allele
+    frequency conditioning.
+
+    FIXME: Drawing multiple mutations using the same mutation type causes
+           erroneous allele frequency calculation in the SLiM code!
```

Comment:
what do you mean here more specifically, @grahamgower?

Reply:
The allele frequency calculation is a bit fragile. I didn't think too hard about it when I originally wrote it, because my use case was limited to drawing only one mutation. What do you think, should we change this `af()` function, or disallow multiple drawn mutations of the same mutation type?
```
// Return the allele frequency of a drawn mutation in the specified population.
// Assumes there's only one mutation of the given type.
function (float$)af(object$ mut_type, object$ pop) {
    mut = sim.mutationsOfType(mut_type);
    if (length(mut) == 0) {
        return 0.0;
    }
    return sim.mutationFrequencies(pop, mut);
}
```